### DISCUSSION TITLE
How to fix GPL-incompatible module zfs.ko uses GPL-only symbol ???

### DISCUSSION THREAD

**Discussion by dipol0** [openzfs/zfs#13471] [Q&A]
I try build kernel 5.17-5 with zfs-master.

Get error:

```
  CC [M]  fs/zfs/zfs/vdev_raidz_math_avx512bw.o
  CC [M]  fs/zfs/zfs/vdev_raidz_math_avx512f.o
  CC [M]  fs/zfs/zfs/vdev_raidz_math_sse2.o
  CC [M]  fs/zfs/zfs/vdev_raidz_math_ssse3.o
  LD [M]  fs/zfs/spl.o
  LD [M]  fs/zfs/zfs.o
  MODPOST modules-only.symvers
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'perf_trace_buf_alloc'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'blk_queue_write_cache'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_begin_mask'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_handle_return'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kvfree_call_rcu'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_buffer_reserve'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_end'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_buffer_commit'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_raw_init'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'bio_associate_blkg'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol '__trace_trigger_soft_disabled'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_reg'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'perf_trace_run_bpf_submit'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'bpf_trace_run4'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'bpf_trace_run3'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'bpf_trace_run1'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'bpf_trace_run2'
make[2]: *** [scripts/Makefile.modpost:134: modules-only.symvers] Ошибка 1
make[2]: *** Удаляется файл «modules-only.symvers»
make[1]: *** [Makefile:1746: modules] Ошибка 2
make[1]: выход из каталога «/home/manager/src/linux-5.17.5»
make: *** [debian/ruleset/targets/common.mk:302: debian/stamp/build/kernel] Ошибка 2
```
**How to fix it ?**

**Comment by jittygitty** [comment]
@dipol0 Can you verify if this an official and standard debian kernel without mods or backports? Also did you build with PREEMPT?
https://github.com/openzfs/zfs/issues/13054
see some comments in there and links to other issues, though referring to 5.16 not sure if resolved already etc.

One fix you can try, is simply building ZFS by telling it that it is GPL, (technically ZFS "binary is GPL" anyway) see:
https://github.com/openzfs/zfs/issues/13415
(sorry I meant this: https://github.com/openzfs/zfs/issues/11357 )
Specifically: 
https://github.com/openzfs/zfs/issues/11357#issuecomment-1115205182
https://github.com/openzfs/zfs/issues/11357#issuecomment-1115486147

**Reply by dipol0** [reply]
@jittygitty i try with and without PREEMPT. I try script

```
grep -r CDDL include/zfs/|grep -v '\*'|grep -v bsd|cut -d':' -f1|while read FL ; do 
  sed -i 's|ZFS_META_LICENSE = CDDL|ZFS_META_LICENSE = GPL|; s|#define ZFS_META_LICENSE "CDDL"|#define ZFS_META_LICENSE "GPL"|' $FL
done
```
but not have result.

**Comment by jittygitty** [comment]
@dipol0 After you changed META in file from CDDL to GPL. Did you re-run ./autogen.sh && ./configure and build zfs again? Please paste what you did and the output/result because this should have fixed error if done correctly. Maybe also paste META file license info to verify. (After you edit META to GPL you must re-build, ie autogen, configure and "make" and install again to overwrite previous version.)
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Building%20ZFS.html
To "verify", when you are done, run the modinfo zfs command and paste the modinfo output so we can see if zfs module reports as GPL.

**Reply by dipol0** [reply]
@jittygitty 

mkdir src
cd sc
wget https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.17.9.tar.xz
tar -xvf linux-5.17.9.tar.x
git clone https://github.com/openzfs/zfs.git
cd linux-5.17.9
yes ''|make defconfig
make prepare
cd ..

Create zfs-mod1.sh script:

```
#!/bin/bash
grep -r CDDL zfs/|grep -v '\*'|grep -v bsd|cut -d':' -f1|while read FL ; do 
  sed -i 's|ZFS_META_LICENSE = CDDL|ZFS_META_LICENSE = GPL|; s|#define ZFS_META_LICENSE "CDDL"|#define ZFS_META_LICENSE "GPL"|' $FL
done
```

run it

./zfs-mod1.sh
cd zfs
cat META

_**Meta:          1
Name:          zfs
Branch:        1.0
Version:       2.1.99
Release:       1
Release-Tags:  relext
License:       CDDL
Author:        OpenZFS
Linux-Maximum: 5.17
Linux-Minimum: 3.10**_

PS ??? (need change "CDDL" to "GPL" manualy?)

sh ./autogen.sh
./configure --prefix=/ --libdir=/lib -includedir=/usr/include --enable-linux-builtin=yes --with-linux=/src/linux-5.17.9 --with-linux-obj=/src/linux-5.17.9
./copy-builtin /src/linux-5.17.9
cd ..
cd linux-5.17.9
make xconfig

Find zfs - chack as module (M) 
Find and set (default) - PREEMPT_VOLUNTARY=y
Save .config

CC="ccache gcc" fakeroot make-kpkg -j$(nproc) --uc --us --initrd --append-to-version=-zfs kernel_image kernel_headers kernel_source

And get errors (see attach file for full log)
[make-kernel.txt](https://github.com/openzfs/zfs/files/8725863/make-kernel.txt)

  LD [M]  fs/zfs/zfs.o
  MODPOST modules-only.symvers
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'blk_queue_write_cache'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_begin_mask'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kvfree_call_rcu'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_end'
make[2]: *** [scripts/Makefile.modpost:134: modules-only.symvers] Ошибка 1
make[2]: *** Удаляется файл «modules-only.symvers»
make[1]: *** [Makefile:1746: modules] Ошибка 2
make[1]: выход из каталога «/home/manager/src/linux-5.17.9»
make: *** [debian/ruleset/targets/common.mk:302: debian/stamp/build/kernel] Ошибка 2

-------------------------------
cat zfs.release

2.1.99-1217_ge02d84d3a

-------------------------------
OK

clean all. and make same but edit META in zfs from CDDL to GPL and sh ./autogen.sh && ./configure...
goto kernel source and make it....and get errors:

....
  LD [M]  fs/zfs/zfs.o
  MODPOST modules-only.symvers
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'perf_trace_buf_alloc'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'blk_queue_write_cache'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_begin_mask'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_handle_return'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kvfree_call_rcu'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_buffer_reserve'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_end'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_buffer_commit'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_raw_init'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol '__trace_trigger_soft_disabled'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_reg'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'perf_trace_run_bpf_submit'
make[2]: *** [scripts/Makefile.modpost:134: modules-only.symvers] Ошибка 1
make[2]: *** Удаляется файл «modules-only.symvers»
make[1]: *** [Makefile:1746: modules] Ошибка 2
make[1]: выход из каталога «/home/manager/src/linux-5.17.9»
make: *** [debian/ruleset/targets/common.mk:302: debian/stamp/build/kernel] Ошибка 2
________________________
Ok. try CONFIG_PREEMPTION_NONE=y get errors:

  LD [M]  fs/zfs/zfs.o
  MODPOST modules-only.symvers
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'perf_trace_buf_alloc'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'blk_queue_write_cache'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_begin_mask'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_handle_return'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kvfree_call_rcu'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_buffer_reserve'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_end'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_buffer_commit'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_raw_init'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol '__trace_trigger_soft_disabled'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'trace_event_reg'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'perf_trace_run_bpf_submit'
make[2]: *** [scripts/Makefile.modpost:134: modules-only.symvers] Ошибка 1
make[2]: *** Удаляется файл «modules-only.symvers»
make[1]: *** [Makefile:1746: modules] Ошибка 2
make[1]: выход из каталога «/home/manager/src/linux-5.17.9»
make: *** [debian/ruleset/targets/common.mk:302: debian/stamp/build/kernel] Ошибка 2

________________________
Previous i build 5.17.5 kernel with zfs-2.1.99-717_g73e972af7 and bee fine. My current kernel 5.17.5. modinfo zfs:

filename:       /lib/modules/5.17.5-zfs/kernel/fs/zfs/zfs/zfs.ko
version:        2.1.99-717_g73e972af7
license:        GPL
author:         OpenZFS
description:    ZFS
alias:          devname:zfs
alias:          char-major-10-249
srcversion:     2EF15F0093CFA7996B0F8ED
depends:        spl,znvpair,icp,zlua,zzstd,zunicode,zcommon,zavl
retpoline:      Y
intree:         Y
name:           zfs
vermagic:       5.17.5-zfs SMP preempt mod_unload modversions 
sig_id:         PKCS#7
signer:         Build time autogenerated kernel key

But kernel 5.17.9 and current zfs (master) 2.1.99-1217_ge02d84d3a dont build.

**Comment by jittygitty** [comment]
@dipol0 If you did finally get it so that compiled "binary" advertises as GPL, then I'm not sure why it would still complain of using GPL-only symbols if in fact the currently loaded module was set as GPL license, I don't have time to test, I hadn't gotten around to testing this cddl->gpl myself yet, but maybe @sempervictus or @Ornias1993 can look and give you some hints, I skimmed this but have no time right now.

**Comment by jittygitty** [comment]
@dipol0 Sorry still busy and no time to duplicate your setup, but if you can't get rid of error even when running zfs.ko binary that says GPL, just click the button to the right that says "create issue from discussion" so it can be dealt with as an actual bug and others can help. https://github.com/openzfs/zfs/issues/new?created_from_discussion_number=13471

**Comment by whatever1983** [comment]
The issue is cropping up again in recent builds.  I am unable to build 2.12, and 2.2rc1 zfs against Linux kernel 6.3 series on Ubuntu MATE 20.04.6.  Last successful built kernel was 6.2.11-xanmod1 with 2.12 zfs.

Ironically, the Ubuntu MATE 23.10 mantic's zfs-dkms package was versioned 2.1.99 and you can dkms the 6.4.3-xanmod1 kernel successfully, even built against 6.4 xanmod kernel.  How ironic.

  AR      built-in.a
  AR      vmlinux.a
  LD      vmlinux.o
  OBJCOPY modules.builtin.modinfo
  GEN     modules.builtin
  GEN     .vmlinux.objs
  MODPOST Module.symvers
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_end'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kvfree_call_rcu'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'kernel_fpu_begin_mask'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'blk_queue_write_cache'
make[5]: *** [scripts/Makefile.modpost:136: Module.symvers] Error 1
make[4]: *** [Makefile:2133: modpost] Error 2
make[3]: *** [Makefile:357: __build_one_by_one] Error 2
make[2]: *** [debian/rules:10: build-arch] Error 2
dpkg-buildpackage: error: debian/rules binary subprocess returned exit status 2
make[1]: *** [scripts/Makefile.package:139: bindeb-pkg] Error 2
make: *** [Makefile:1812: bindeb-pkg] Error 2

**Reply by rincebrain** [reply]
Please open your own issue, putting every case of "GPL-only symbol" in one issue is going to lead to a lot of people replying that the solution to one of these didn't fix their problem, and so on.
